Mathlib/Data/Nat/Digits/Lemmas.lean

```diff
@@ -137,6 +139,116 @@ theorem base_pow_length_digits_le (b m : ℕ) (hb : 1 < b) :
   rcases b with (_ | _ | b) <;> try simp_all
   exact base_pow_length_digits_le' b m
 
+/-- The sum of the digits of a number less than its base is the number itself.
+-/
+@[simp]
+theorem digits_sum_eq_self {x b : ℕ} (hxb : x < b) : (b.digits x).sum = x := by
+  by_cases x_eq : x = 0
+  · simp [x_eq]
+  simp [Nat.digits_of_lt b x x_eq hxb]
+
+/-- Multiplying by the base does not change the digit sum.
+-/
+@[simp]
+theorem digits_sum_base_mul_cancel {n b : ℕ} (pos_b : 0 < b) :
+    (b.digits (b * n)).sum = (b.digits n).sum := by
+  obtain _ | n := n
+  · simp
+  by_cases hb : b = 1
+  · simp [hb]
+  push_neg at hb
+  conv_lhs => rhs; rhs; rw [<- pow_one b]
+  rw [Nat.digits_base_pow_mul (Nat.lt_of_le_of_ne pos_b hb.symm)
+    (zero_lt_succ n)]
+  simp
+
+/-- Closed form formula for summing digit sums up to a power of the base. This
+is equivalent to applying Gauss's formula across digit columns.
+-/
+theorem sum_digit_sum_base_pow_eq {k b : ℕ} :
+    2 * ∑ n ∈ Finset.range (b^k), (b.digits n).sum = (b-1) * k * b^k := by
+  by_cases k0 : k = 0
+  · simp [mul_comm, k0]
+  by_cases hb : b = 0 ∨ b = 1
+  · rcases hb with (h | h) <;> simp [h, k0]
+  replace hb : 1 < b := by omega
+  have sum_sum_digits : 2 * ∑ x ∈ Finset.range b, (b.digits x).sum = b * (b - 1) := by
+    rw [mul_comm, <- Finset.sum_range_id_mul_two]
+    congr! with x x_in
+    simp at x_in
+    simp [digits_sum_eq_self x_in]
+  induction' k, (by omega : 1 ≤ k) using Nat.le_induction with m pos_m ih
+  · simp [sum_sum_digits, mul_comm]
+  · specialize ih (Nat.ne_zero_of_lt pos_m)
+    have digits_eq_cons := Nat.digits_eq_cons_digits_div
+      hb (by positivity)
+    have pos_n : ∀ n ∈ Finset.Ico 1 (b^m), 0 < n := by
+      intro n n_in
+      simp at n_in
+      exact n_in.left
+    have sum_pow_add_eq_sum_pair {f : ℕ → ℕ} {i j b : ℕ} :
+      ∑ x ∈ Finset.range (b^(i+j)), f x =
+      ∑ p ∈ Finset.range (b^i) ×ˢ Finset.range (b^j),
+        f (b^j * p.fst + p.snd) := by
```

Comment:
When we say "generalise this to theorem A", we don't mean "put A inside a `have` somewhere in a big proof where nobody can use it"; we mean "state A separately as a separate lemma so that people can use it in the future".

Reply:
I misunderstood that those lemmas were substantial enough to warrant a stand-alone version.

I made a new PR for them [here](https://github.com/leanprover-community/mathlib4/pull/27297), and I will update this PR later to depend on it (and with `sum_digit_sum_base_pow_eq` broken up).